nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,200 @@
+        return false;
+      }
+    }
+    geometry_msgs::msg::PolygonStamped poly_out, polygon_stamped;
+    geometry_msgs::msg::TransformStamped tf = tf2::toMsg(tf_transform);
+    polygon_stamped.header = polygon_instance.header;
+    polygon_stamped.polygon = polygon_instance.polygon.polygon;
+    tf2::doTransform(polygon_stamped, poly_out, tf);
+    convertPolygonStampedToPoints(poly_out, data);
+  }
+  return true;
+}
+
+void PolygonSource::convertPolygonStampedToPoints(
+  const geometry_msgs::msg::PolygonStamped & polygon, std::vector<Point> & data) const
+{
+  /*  Full function generated by GPT */
+
+  // Iterate over the vertices of the polygon
+  for (size_t i = 0; i < polygon.polygon.points.size(); ++i) {
+    const auto & current_point = polygon.polygon.points[i];
+    const auto & next_point =
+      polygon.polygon.points[(i + 1) % polygon.polygon.points.size()];
+
+    // Calculate the distance between the current and next points
+    double segment_length =
+      std::hypot(next_point.x - current_point.x, next_point.y - current_point.y);
+
+    // Calculate the number of points to sample in the current segment
+    int num_points_in_segment =
+      std::max(static_cast<int>(std::ceil(segment_length / sampling_distance_)), 1);
+
+    // Calculate the step size for each pair of vertices
+    const double dx = (next_point.x - current_point.x) / num_points_in_segment;
+    const double dy = (next_point.y - current_point.y) / num_points_in_segment;
+
+    // Sample the points with equal spacing
+    for (int j = 0; j <= num_points_in_segment; ++j) {
+      Point p;
+      p.x = current_point.x + j * dx;
+      p.y = current_point.y + j * dy;
+      data.push_back(p);
+    }
+  }
+}
+
+void PolygonSource::getParameters(std::string & source_topic)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  getCommonParameters(source_topic);
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".sampling_distance", rclcpp::ParameterValue(0.1));
+  sampling_distance_ = node->get_parameter(source_name_ + ".sampling_distance").as_double();
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".polygon_similarity_threshold", rclcpp::ParameterValue(2.0));
+  polygon_similarity_threshold_ =
+    node->get_parameter(source_name_ + ".polygon_similarity_threshold").as_double();
+}
+
+void PolygonSource::dataCallback(geometry_msgs::msg::PolygonInstanceStamped::ConstSharedPtr msg)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+  auto curr_time = node->now();
+
+  // check if older similar polygon exists already and replace it with the new one
+  for (auto & polygon_stamped : data_) {
+    if (msg->header.frame_id == polygon_stamped.header.frame_id &&
+      rclcpp::Time(msg->header.stamp) >= rclcpp::Time(polygon_stamped.header.stamp) &&
+      areSamePolygons(*msg, polygon_stamped))
```

Comment:
I think the frame_ids don't necessarily have to be the same for the polygons to be considered the same and hence be replaced. It would be weird but if a user decides to start publishing the polygon with the same id with a different frame_id... why not?

For the time we could maybe apply the same reasoning but it's a bit more delicate because it's used for removing the polygons by timeout.

Reply:
Fixed